io/src/lib.rs

```diff
@@ -0,0 +1,504 @@
+    use core::fmt::{Debug, Display, Formatter};
+    #[cfg(any(feature = "alloc", feature = "std"))]
+    use alloc::boxed::Box;
+
+    #[cfg(all(feature = "alloc", not(feature = "std")))]
+    mod sealed {
+        use alloc::boxed::Box;
+        use alloc::string::String;
+        use core::fmt::Debug;
+        pub trait IntoBoxDynDebug {
+            fn into(self) -> Box<dyn Debug + Send + Sync + 'static>;
+        }
+        impl IntoBoxDynDebug for &str {
+            fn into(self) -> Box<dyn Debug + Send + Sync + 'static> {
+                Box::new(String::from(self))
+            }
+        }
+        impl IntoBoxDynDebug for String {
+            fn into(self) -> Box<dyn Debug + Send + Sync + 'static> {
+                Box::new(self)
+            }
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct Error {
+        kind: ErrorKind,
+
+        #[cfg(feature = "std")]
+        error: Option<Box<dyn std::error::Error + Send + Sync + 'static>>,
+        #[cfg(all(feature = "alloc", not(feature = "std")))]
+        error: Option<Box<dyn Debug + Send + Sync + 'static>>,
+    }
+    impl Error {
+        #[cfg(feature = "std")]
+        pub fn new<E>(kind: ErrorKind, error: E) -> Error
+        where
+            E: Into<Box<dyn std::error::Error + Send + Sync + 'static>>,
+        {
+            Self { kind, error: Some(error.into()) }
+        }
+        #[cfg(all(feature = "alloc", not(feature = "std")))]
+        pub fn new<E: sealed::IntoBoxDynDebug>(kind: ErrorKind, error: E) -> Error {
+            Self { kind, error: Some(error.into()) }
+        }
+
+        pub fn kind(&self) -> ErrorKind {
+            self.kind
+        }
+    }
+
+    impl From<ErrorKind> for Error {
+        fn from(kind: ErrorKind) -> Error {
+            Self {
+                kind,
+                #[cfg(any(feature = "std", feature = "alloc"))]
+                error: None,
+            }
+        }
+    }
+
+    impl Display for Error {
+        fn fmt(&self, fmt: &mut Formatter) -> core::result::Result<(), core::fmt::Error> {
+            fmt.write_fmt(format_args!("I/O Error: {}", self.kind.description()))?;
+            #[cfg(any(feature = "alloc", feature = "std"))]
+            if let Some(e) = &self.error {
+                fmt.write_fmt(format_args!(". {:?}", e))?;
+            }
+            Ok(())
+        }
+    }
+
+    #[cfg(feature = "std")]
+    impl std::error::Error for Error {
+        fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+            self.error.as_ref().and_then(|e| e.as_ref().source())
+        }
+        #[allow(deprecated)]
+        fn description(&self) -> &str {
+            match self.error.as_ref() {
+                Some(e) => e.description(),
+                None => self.kind.description(),
+            }
+        }
+        #[allow(deprecated)]
+        fn cause(&self) -> Option<&dyn std::error::Error> {
+            self.error.as_ref().and_then(|e| e.as_ref().cause())
+        }
```

Comment:
Out of interest, what is the motivation for implementing these deprecated trait methods? (deprecated in Rust 1.33)